agent/lib_aws_sdk_php.c

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2024 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * Functions relating to instrumenting the AWS-SDK-PHP.
+ * https://github.com/aws/aws-sdk-php
+ */
+#include "php_agent.h"
+#include "php_wrapper.h"
+#include "fw_hooks.h"
+#include "fw_support.h"
+#include "util_logging.h"
+#include "php_call.h"
+#include "lib_aws_sdk_php.h"
+
+#define PHP_PACKAGE_NAME "aws/aws-sdk-php"
+#define PHP_AWS_CLASS_PREFIX "Supportability/library/aws/aws-sdk-php/"
+
+/*
+ * In a normal course of events, the following line will always work
+ * zend_eval_string("Aws\\Sdk::VERSION;", &retval, "Get AWS Version")
+ * By the time we have detected the existence of the aws-sdk-php and with
+ * default composer profject settings, it callable even from
+ * nr_aws_sdk_php_enable which will automatically load the class if it isn't
+ * loaded yet and then evaluate the string. However, in the rare case that files
+ * are not loaded via autoloader and/or have non-default composer classload
+ * settings, if the class is not found, PHP 8.2+ will generate a fatal
+ * unrecoverable uncatchable error error whenever it cannot find a class. While
+ * calling this from nr_aws_sdk_php_enable would have been great and would allow
+ * the sdk version value to be set only once, to avoid the very unlikely but not
+ * impossible fatal error, this will be called from the
+ * "Aws\\ClientResolver::_apply_user_agent" wrapper which GUARANTEES that
+ * aws/sdk exists and is already loaded.
+ *
+ *
+ * Additionally given that aws-sdk-php is currently detected from the
+ * AwsClient.php file, this method will always be called when a client is
+ * created unlike Sdk::construct which doesn't show with PHP 8.2+.
+ *
+ * Using Aws/Sdk::__construct for version is currently nonviable as it is
+ * unreliable as a version determiner.
+ * Having separate functionality to extract from Aws/Sdk::__construct
+ * is both not required and is redundant and causes additional overhead and
+ * so only one function is needed to extract version.
+ *
+ * Aws\\ClientResolver::_apply_user_agent a reliable function as it is
+ * always called on client initialization since it is key to populating
+ * the request headers, and it loads Sdk by default.
+ *
+ * Concerns about future/past proofing to the checking prioritized the following
+ * implementation vs using the eval method.
+ */
+extern void lib_aws_sdk_php_handle_version() {
+  zval* zval_version = NULL;
+  zend_class_entry* class_entry = NULL;
+  char* version = NULL;
+
+  class_entry = nr_php_find_class("aws\\sdk");
```

Comment:
On PHP 7.4, when using this sample code to construct `Aws\S3\S3Client`:
```
$s3Client = new Aws\S3\S3Client([
    'profile' => 'default',
    'region' => 'us-west-2'
]);
```
`class_entry` returned here is `NULL`. It is not `NULL` on PHP 8.2. I did not test with other PHPs.

Reply:
Further investigation reveals that on PHP 7.4 `aws/aws-sdk-php/src/Sdk.php` file with the definition of `\Aws\Sdk::Version` constant is loaded **after** `Aws\ClientResolver::_apply_user_agent` executes and that would explain why `nr_php_find_class("aws\\sdk");` returns `NULL`.

Reply:
Great catch, please see: [c27a257](https://github.com/newrelic/newrelic-php-agent/pull/932/commits/c27a25719af98a9ba6e17638e33636b08cc9f670)


NR_PHP_WRAPPER_CALL needs to happen before checking the version number for PHP 7.x.
Since Aws\ClientResolver::_apply_user_agent uses Sdk::VERSION it should be loaded by the time it finishes.

Also updated the multiverse test_php_packages.php test to only call the client constructor and to have no reference or use of the Aws/Sdk class.